src/zernike_fit.jl

```diff
@@ -1,63 +1,109 @@
 # Fitting function zernike_img_fit()
-# Loss function zernikeloss()
-# Gradient function zernikegrad!()
+# Phase diversity Loss function zernikeloss()
+# Phase diversity Gradient function zernikegrad!()
+# PSF based Loss function psfloss()
+# PSF based Gradient function psfgrad!()
 # Matrix preparation function loss_prep()
 #
 
-function loss_prep(Z, img, Hz, Zval)
-    imsz = size(img)
-    Hk, Sk = ZernFT(Z, Hz, Zval, imsz)
+function loss_prep(Z, img, Hz, Zval, Z_de)
+    imsz = size(img)[1:3]
+    K = length(size(img)) == 4 ? size(img)[4] : 1
+    Hk = zeros(Complex{Float64}, (imsz..., K))
+    Sk = zeros(Complex{Float64}, (imsz..., K))
+    for i = 1:K
+        Zk = Z .+ Z_de[i]
+        Hk[:, :, :, i], Sk[:, :, :, i] = ZernFT(Zk, Hz, Zval, imsz)
+    end
 
     #compute transform of image
-    Dk = fft(img)
+    Dk = zeros(Complex{Float64}, (imsz..., K))
+    for i = 1:K
+        Dk[:, :, :, i] = fft(img[:, :, :, i])
+    end
 
     # penalty prep
-    S2tot = abs2.(Sk)
+
+    S2tot = dropdims(sum(abs2.(Sk), dims = 4), dims = 4)
     ukeep = S2tot .> eps()
-    D2tot = abs2.(Dk)
-    DdotS = Dk .* conj(Sk)
+    D2tot = dropdims(sum(abs2.(Dk), dims = 4), dims = 4)
+    DdotS = dropdims(sum(Dk .* conj(Sk), dims = 4), dims = 4)
     return Hk, Dk, Sk, S2tot, ukeep, D2tot, DdotS
 end
 
-function zernikeloss(Z, img, Hz, Zval)
-    _, _, _, S2tot, ukeep, D2tot, DdotS = loss_prep(Z, img, Hz, Zval)
+# Phase diversity loss and gradient function
+function zernikeloss(Z, img, Hz, Zval, Z_de)
+    _, _, _, S2tot, ukeep, D2tot, DdotS = loss_prep(Z, img, Hz, Zval, Z_de)
     num = abs2.(DdotS)
 
-    # F = zeros(Complex{Float64}, size(DdotS))
-    # for id in findall(ukeep)
-    #     F[id] = DdotS[id] / S2tot[id]
-    # end
-
-
     return -sum(num[ukeep] ./ S2tot[ukeep]) + sum(D2tot)
 end
 
-function zernikegrad!(g, Z, img, Hz, Zval)
-    Hk, Dk, Sk, S2tot, ukeep, _, DdotS = loss_prep(Z, img, Hz, Zval)
+function zernikegrad!(g, Z, img, Hz, Zval, Z_de)
+    Hk, Dk, Sk, S2tot, ukeep, _, DdotS = loss_prep(Z, img, Hz, Zval, Z_de)
 
+    K = size(img)[4]
     coef1 = S2tot .* DdotS
     coef2 = DdotS .* conj(DdotS)
-    grad_num = coef1 .* conj(Dk) - coef2 .* conj(Sk)
-    Zk = zeros(Complex{Float64}, size(grad_num))
-    for id in findall(ukeep)
-        Zk[id] = grad_num[id] ./ (S2tot[id] .^ 2)
+    grad_num = zeros(Complex{Float64}, size(img)[1:3])
+    Zk = zeros(Complex{Float64}, size(img))
+    ZconvH = zeros(Complex{Float64}, size(img))
+    for i = 1:K
+        grad_num[:, :, :] = coef1 .* conj(Dk[:, :, :, i]) - coef2 .* conj(Sk[:, :, :, i])
+        for id in findall(ukeep)
```

Comment:
If `ukeep` doesn't change on each iteration, it would be best to compute `findall(ukeep)` once, outside the loop, instead of on each iteration.